### ISSUE TITLE
Inverse Transformation of the Data while Plotting Ground Truth Vs Prediction 

### ISSUE THREAD

**Issue by vinayakrajurs** [thuml/Autoformer#6]
Good Day,
I'm currently working on the forecasting of Univariant time series using your Autoformer in google colab.
I've attached the dataset I'm using below and I am currently trying to predict the Voltage data in the dataset.

[dataset4.csv](https://github.com/thuml/Autoformer/files/7655527/dataset4.csv)

After the Prediction of the test data, the prediction and ground truth values are plotted however they are not scaled back to their original values.
I want your help in doing this.

I've made the some changes in the run.py and data_factory.py to fit my custom dataset in the code below 

[run.py.pdf](https://github.com/thuml/Autoformer/files/7655545/run.py.pdf) 
- Line 69 I've added a new line of code: `parser.add_argument('--inverse', action='store_true', help='inverse output data', default=False)`
and made some changes with the epochs and enc_in , dec_in and c_out values.

[data_factory.py.pdf](https://github.com/thuml/Autoformer/files/7655552/data_factory.py.pdf) 
- Line 9 I've inserted dataset4 as the custom data set.

After running the code : 
`!python run.py --is_training 1 --model_id test --model Informer --data dataset4`

I was able to train the model and run it on test data and got the corresponding MAE and MSE values

This is the code used to Plot the results:

`%cd /content/drive/MyDrive/Autoformer_main/Autoformer-main/results/test_Informer_dataset4_ftM_sl96_ll48_pl24_dm512_nh8_el2_dl1_df2048_fc5_ebtimeF_dtTrue_test_0
!ls`

`import numpy as np
preds = np.load('pred.npy')
trues = np.load('true.npy')
import matplotlib.pyplot as plt
import seaborn as sns`

`import matplotlib.pyplot as plt
plt.figure()
plt.plot (trues[:,1,-1], label='GroundTruth')
plt.plot (preds[:,1,-1], label='Prediction')
plt.legend()
from google.colab import files
plt.show()`

In the Plot below you can see that the ground truth value of Voltage is between -2 and 1:

![image](https://user-images.githubusercontent.com/88593317/144740643-bf207d27-73cb-4958-a3d8-752e9c975d7a.png)

but if you look at the original dataset, I've plotted the ground Truth value of voltage and its is between 3 and 4.5 as shown below:

![image](https://user-images.githubusercontent.com/88593317/144744284-5df7fd24-8fcc-4d55-a6e0-ea074030a504.png)

You are using Standard scaler to scale the data, however I am unable to re-scale the standardized data back to its original values while plotting both the ground truth and prediction value.

Please help in using the inverse transform to re-scale my data and predictions back to its original values.

Thank you

**Comment by wuhaixu2016** [comment]
Hi, Thanks for your attention.
(1) I know this problem about without re-scale. But sadly, this problem appears in Informer and a branch of follow-up baselines. To conduct a fair comparison, we also calculate the MSE and MAE without re-scale.

(2) You can re-scale the data by the Mean and Std recorded by the Standard scaler. I will update the `--inverse` argument soon.

**Comment by vinayakrajurs** [comment]
hello, thank you for the quick reply, 
I'm able to rescale the values back in the Informer.
In Autoformer even with the new argument `--inverse` , I'm still not able to get the values back, If i try adding `--inverse true` in the main `!python run.py --is_training 1 --model_id test --model Informer --data dataset4` code cell, its showing a syntax error.

**Comment by vinayakrajurs** [comment]
How do I re-scale the data by the Mean and Std recorded by the Standard scaler for Autoformer ?

Thank you

**Comment by vinayakrajurs** [comment]
> Hi, Thanks for your attention. (1) I know this problem about without re-scale. But sadly, this problem appears in Informer and a branch of follow-up baselines. To conduct a fair comparison, we also calculate the MSE and MAE without re-scale.
> 
> (2) You can re-scale the data by the Mean and Std recorded by the Standard scaler. I will update the `--inverse` argument soon.

Thank You, If i wanted the Ground Truth Vs prediction Plot in between (0,1) on the Y-Axis, How should i change the Standard Scale Class?

**Comment by wuhaixu2016** [comment]
> 

(1) In this `exp_main.py`
add `self.Data = data_set` to `_get_data` function.
(2) In the `train`, `vali`, and `test`. add the following code to the prediction results.
mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_
batch_y=batch_y*std_X+mean_X
outputs=outputs*std_X+mean_X

**Comment by wuhaixu2016** [comment]
> > Hi, Thanks for your attention. (1) I know this problem about without re-scale. But sadly, this problem appears in Informer and a branch of follow-up baselines. To conduct a fair comparison, we also calculate the MSE and MAE without re-scale.
> > (2) You can re-scale the data by the Mean and Std recorded by the Standard scaler. I will update the `--inverse` argument soon.
> 
> Thank You, If i wanted the Ground Truth Vs prediction Plot in between (0,1) on the Y-Axis, How should i change the Standard Scale Class?

Hi,

**Comment by wuhaixu2016** [comment]
Sorry for the 'close issue' mistake operation.
--------------------------
About this question 'wanted the Ground Truth Vs prediction Plot in between (0,1)'
(1) The `Standard Scale` will change the data into a standard Gaussian distribution (https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html), that is why you can get the value not in the interval (0,1)
(2) If you want to transform the data into (0,1), you can directly divide the max value of the train set. And in this context, you use the uniform prior to the data.

In summary, the `Standard Scale` adopts the `Gaussian prior`, the situation you mentioned will adopt the `uniform prior`.

**Comment by vinayakrajurs** [comment]
> ## Sorry for the 'close issue' mistake operation.
> About this question 'wanted the Ground Truth Vs prediction Plot in between (0,1)' (1) The `Standard Scale` will change the data into a standard Gaussian distribution (https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html), that is why you can get the value not in the interval (0,1) (2) If you want to transform the data into (0,1), you can directly divide the max value of the train set. And in this context, you use the uniform prior to the data.
> 
> In summary, the `Standard Scale` adopts the `Gaussian prior`, the situation you mentioned will adopt the `uniform prior`.

Thank You for the explanation

**Comment by wuhaixu2016** [comment]
> > ## How to re-scale the data
> > (1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function. (2) In the `train`, `vali`, and `test`. add the following code to the prediction results. mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y_std_X+mean_X outputs=outputs_std_X+mean_X
> 
> Thank you for the quick reply, In (2) could you please specify the line of code for `test` , `vali` and `test` for ease, could you attach `exp_main.py` file with above codes already implemented ? Sorry for the Inconvenience, I'm really new at this and I'm not able to make the changes to `exp_main.py` as i don't know where to include : `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` in `exp_main.py`
> 
> Thank you

`mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` can be add to line39-line40
`batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` can be add to line81 of vali, line229 of test. As for the train, you don't need to change if your only want to get a "real-scale" prediction.

**Comment by vinayakrajurs** [comment]
> > > ## How to re-scale the data
> > > (1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function. (2) In the `train`, `vali`, and `test`. add the following code to the prediction results. mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y_std_X+mean_X outputs=outputs_std_X+mean_X
> > 
> > 
> > Thank you for the quick reply, In (2) could you please specify the line of code for `test` , `vali` and `test` for ease, could you attach `exp_main.py` file with above codes already implemented ? Sorry for the Inconvenience, I'm really new at this and I'm not able to make the changes to `exp_main.py` as i don't know where to include : `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` in `exp_main.py`
> > Thank you
> 
> `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` can be add to line39-line40 `batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` can be add to line81 of vali, line229 of test. As for the train, you don't need to change if your only want to get a "real-scale" prediction.

I've made the the above changes as u suggested however when im running t he main python code `!python run.py --is_training 1 --model_id test --model Autoformer --data dataset4`  I'm getting the following error :

> Args in experiment:
Namespace(activation='gelu', batch_size=32, c_out=3, checkpoints='./checkpoints/', d_ff=2048, d_layers=1, d_model=512, data='dataset4', data_path='dataset4.csv', dec_in=3, des='test', devices='0,1,2,3', distil=True, do_predict=False, dropout=0.05, e_layers=2, embed='timeF', enc_in=3, factor=5, features='M', freq='s', gpu=0, inverse=True, is_training=1, itr=2, label_len=48, learning_rate=0.0001, loss='mse', lradj='type1', model='Autoformer', model_id='test', moving_avg=25, n_heads=8, num_workers=0, output_attention=False, patience=10, pred_len=24, root_path='./data/ETT', seq_len=96, target='Voltage', train_epochs=2, use_amp=False, use_gpu=True, use_multi_gpu=False)
Use GPU: cuda:0
>>>>>>>start training : test_Autoformer_dataset4_ftM_sl96_ll48_pl24_dm512_nh8_el2_dl1_df2048_fc5_ebtimeF_dtTrue_test_0>>>>>>>>>>>>>>>>>>>>>>>>>>
Traceback (most recent call last):
  File "run.py", line 115, in <module>
    exp.train(setting)
  File "/content/drive/My Drive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 92, in train
    train_data, train_loader = self._get_data(flag='train')
  File "/content/drive/My Drive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 39, in _get_data
    mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_
AttributeError: 'Exp_Main' object has no attribute 'Data'

I've attached a copy of exp_main.py that im using below :
[exp_main.py.pdf](https://github.com/thuml/Autoformer/files/7662713/exp_main.py.pdf)

Please help, 
Thank you

**Comment by vinayakrajurs** [comment]
> > > ## How to re-scale the data
> > > (1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function. (2) In the `train`, `vali`, and `test`. add the following code to the prediction results. mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y_std_X+mean_X outputs=outputs_std_X+mean_X
> > 
> > 
> > Thank you for the quick reply, In (2) could you please specify the line of code for `test` , `vali` and `test` for ease, could you attach `exp_main.py` file with above codes already implemented ? Sorry for the Inconvenience, I'm really new at this and I'm not able to make the changes to `exp_main.py` as i don't know where to include : `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` in `exp_main.py`
> > Thank you
> 
> `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` can be add to line39-line40 `batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` can be add to line81 of vali, line229 of test. As for the train, you don't need to change if your only want to get a "real-scale" prediction.

Also if i try running the code *without* adding `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` to line39-line40 in def _get_data in `exp_main.py`
 I'm Getting the following error : 
`File "run.py", line 115, in <module>
    exp.train(setting)
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 170, in train
    vali_loss = self.vali(vali_data, vali_loader, criterion)
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 81, in vali
    batch_y=batch_y*std_X+mean_X 
NameError: name 'std_X' is not defined`

**Comment by wuhaixu2016** [comment]
> ## How to re-scale the data
> (1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function. (2) In the `train`, `vali`, and `test`. add the following code to the prediction results. mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X

Recall this:
(1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function.

**Comment by vinayakrajurs** [comment]
> > > ## How to re-scale the data
> > > (1) In this `exp_main.py` add `self.Data = data_set` to `_get_data` function. (2) In the `train`, `vali`, and `test`. add the following code to the prediction results. mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y_std_X+mean_X outputs=outputs_std_X+mean_X
> > 
> > 
> > Thank you for the quick reply, In (2) could you please specify the line of code for `test` , `vali` and `test` for ease, could you attach `exp_main.py` file with above codes already implemented ? Sorry for the Inconvenience, I'm really new at this and I'm not able to make the changes to `exp_main.py` as i don't know where to include : `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_ batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` in `exp_main.py`
> > Thank you
> 
> `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` can be add to line39-line40 `batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` can be add to line81 of vali, line229 of test. As for the train, you don't need to change if your only want to get a "real-scale" prediction.

(1) In `exp_main.py` I've added `self.Data = data_set` line 41 and `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` Line 40 to `_get_data` class 
and I've added `batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` in `def vali` Line 83-84 and as well as
in `def test` Line 232-233
This is the error that I'm getting:

`!python run.py --is_training 1 --model_id test --model Autoformer --data dataset4
Args in experiment:
Namespace(activation='gelu', batch_size=32, c_out=3, checkpoints='./checkpoints/', d_ff=2048, d_layers=1, d_model=512, data='dataset4', data_path='dataset4.csv', dec_in=3, des='test', devices='0,1,2,3', distil=True, do_predict=False, dropout=0.05, e_layers=2, embed='timeF', enc_in=3, factor=5, features='M', freq='s', gpu=0, inverse=True, is_training=1, itr=2, label_len=48, learning_rate=0.0001, loss='mse', lradj='type1', model='Autoformer', model_id='test', moving_avg=25, n_heads=8, num_workers=0, output_attention=False, patience=10, pred_len=24, root_path='./data/ETT', seq_len=96, target='Voltage', train_epochs=2, use_amp=False, use_gpu=True, use_multi_gpu=False)
Use GPU: cuda:0
>>>>>>>start training : test_Autoformer_dataset4_ftM_sl96_ll48_pl24_dm512_nh8_el2_dl1_df2048_fc5_ebtimeF_dtTrue_test_0>>>>>>>>>>>>>>>>>>>>>>>>>>
train 2771
Traceback (most recent call last):
  File "run.py", line 115, in <module>
    exp.train(setting)
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 93, in train
    train_data, train_loader = self._get_data(flag='train')
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 40, in _get_data
    mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_
AttributeError: 'Exp_Main' object has no attribute 'Data'`

![image](https://user-images.githubusercontent.com/88593317/145046842-69ff8fd3-27ce-48fc-8346-ded7eaba99ab.png)

(2) I've also tried adding  `mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_` in `vali` and `test` before `batch_y=batch_y*std_X+mean_X outputs=outputs*std_X+mean_X` line 
The error I'm getting is:

`Args in experiment:
Namespace(activation='gelu', batch_size=32, c_out=3, checkpoints='./checkpoints/', d_ff=2048, d_layers=1, d_model=512, data='dataset4', data_path='dataset4.csv', dec_in=3, des='test', devices='0,1,2,3', distil=True, do_predict=False, dropout=0.05, e_layers=2, embed='timeF', enc_in=3, factor=5, features='M', freq='s', gpu=0, inverse=True, is_training=1, itr=2, label_len=48, learning_rate=0.0001, loss='mse', lradj='type1', model='Autoformer', model_id='test', moving_avg=25, n_heads=8, num_workers=0, output_attention=False, patience=10, pred_len=24, root_path='./data/ETT', seq_len=96, target='Voltage', train_epochs=2, use_amp=False, use_gpu=True, use_multi_gpu=False)
Use GPU: cuda:0
>>>>>>>start training : test_Autoformer_dataset4_ftM_sl96_ll48_pl24_dm512_nh8_el2_dl1_df2048_fc5_ebtimeF_dtTrue_test_0>>>>>>>>>>>>>>>>>>>>>>>>>>
train 2771
val 391
test 802
Epoch: 1 cost time: 45.077964782714844
Traceback (most recent call last):
  File "run.py", line 115, in <module>
    exp.train(setting)
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 172, in train
    vali_loss = self.vali(vali_data, vali_loader, criterion)
  File "/content/drive/MyDrive/Autoformer_main/Autoformer-main/exp/exp_main.py", line 82, in vali
    mean_X, std_X = self.Data.scaler.mean, self.Data_.scaler.scale_
AttributeError: 'StandardScaler' object has no attribute 'mean'`

![image](https://user-images.githubusercontent.com/88593317/145048276-db954c06-38b4-45e6-b980-ece2a3279939.png)

**Comment by vinayakrajurs** [comment]
I've also previously used your Informer model as well for similar forecasting:

![image](https://user-images.githubusercontent.com/88593317/145048640-8b4dfc5c-e7da-4825-b767-769ea03b38b6.png)
The above Image represents the Ground truth Vs prediction for Voltage data from dataset4 after scaling it back to original

I wish to do the same with Autoformer.

**Comment by vinayakrajurs** [comment]
Is there any way to rescale the `pred.npy` and `true.npy` values in results back to the original scale?

**Comment by wuhaixu2016** [comment]
> Is there any way to rescale the `pred.npy` and `true.npy` values in results back to the original scale?

You can re-statistic the mean and std in Train dataset

**Comment by wuhaixu2016** [comment]
> StandardScaler

Can you refer to the standard API? I think this is very simple and very convenient.
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

**Comment by vinayakrajurs** [comment]
> > StandardScaler
> 
> Can you refer to the standard API? I think this is very simple and very convenient. https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
> 
> I **do not** think I can provide more details than this API doc. So just read this doc!

I'll go through the same,
Thank you
Sorry for the inconvenience

**Comment by linbingkong** [comment]
@vinayakrajurs 
[exp_main.pdf](https://github.com/thuml/Autoformer/files/7779819/exp_main.pdf)
you can try this code

**Comment by vinayakrajurs** [comment]
> @vinayakrajurs [exp_main.pdf](https://github.com/thuml/Autoformer/files/7779819/exp_main.pdf) you can try this code

Thank you so much @linbingkong , the code works fine and I'm able to get the denormalized Plots (inverse)

**Comment by white-zhou** [comment]
![image](https://user-images.githubusercontent.com/77036981/230622855-de7d9397-44e4-4713-a6c8-342b41f948c5.png)
我的模型是FEDformer，按照您的方法改了之后出现了这个问题，该如何解决呢？

**Comment by Med-Rokaimi** [comment]
It is not enough to just add the line:
parser.add_argument('--inverse', action='store_true', help='inverse output data', default=True)
to get the test data denormalized in the files preds,npy and trues.npy?

**Comment by Med-Rokaimi** [comment]
> @vinayakrajurs [exp_main.pdf](https://github.com/thuml/Autoformer/files/7779819/exp_main.pdf) you can try this code

Thank you for this example. 
I get this error please:
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

**Comment by Monkcser** [comment]
> @vinayakrajurs [exp_main.pdf](https://github.com/thuml/Autoformer/files/7779819/exp_main.pdf) you can try this code

Thank you for your pdf, it`s very useful!
 And what should I do to obtain a comparison chart between the overall predicted and groundtruth values of the test set?
